pkg/core/disk.go

```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
 	Percent           string `json:"percent"`
 	PayloadProcessNum uint8  `json:"payload_process_num"`
 
-	FillByFallocate bool `json:"fill_by_fallocate"`
+	FillByFAllocate bool `json:"fill_by_fallocate"`
 }
 
-var _ AttackConfig = &DiskOption{}
+func NewDiskOption() *DiskOption {
+	return &DiskOption{
+		CommonAttackConfig: CommonAttackConfig{
+			Kind: DiskAttack,
+		},
+	}
+}
+
+func (opt *DiskOption) PreProcess() (*DiskAttackConfig, error) {
+	if err := opt.CommonAttackConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	path, err := initPath(opt)
+	if err != nil {
+		return nil, err
+	}
 
-func (d *DiskOption) Validate() error {
-	if err := d.CommonAttackConfig.Validate(); err != nil {
-		return err
+	byteSize, err := initSize(opt)
+	if err != nil {
+		return nil, err
 	}
-	var byteSize uint64
-	var err error
-	if d.Size == "" {
-		if d.Percent == "" {
-			return fmt.Errorf("one of percent and size must not be empty, DiskOption : %v", d)
+
+	if opt.Action == DiskFillAction && opt.FillByFAllocate && byteSize != 0 {
+		return &DiskAttackConfig{
+			CommonAttackConfig: opt.CommonAttackConfig,
+			DdOptions:          nil,
+			FAllocateOption: &FAllocateOption{
+				LengthOpt: "-l",
+				Length:    strconv.FormatUint(byteSize, 10),
+				FileName:  path,
+			},
+			Path: path,
+		}, nil
+	}
+
+	ddOptions, err := initDdOptions(opt, path, byteSize)
+	if err != nil {
+		return nil, err
+	}
+	return &DiskAttackConfig{
+		CommonAttackConfig: opt.CommonAttackConfig,
+		DdOptions:          &ddOptions,
+		FAllocateOption:    nil,
+		Path:               path,
+	}, nil
+}
+
+func initDdOptions(opt *DiskOption, path string, byteSize uint64) ([]DdOption, error) {
+	ddBlocks, err := utils.SplitBytesByProcessNum(byteSize, opt.PayloadProcessNum)
+	if err != nil {
+		log.Error("fail to split disk size", zap.Error(err))
+		return nil, err
+	}
+	var ddOpts []DdOption
+	switch opt.Action {
+	case DiskFillAction:
+		for _, block := range ddBlocks {
+			ddOpts = append(ddOpts, DdOption{
+				ReadPath:  "/dev/zero",
+				WritePath: path,
+				BlockSize: block.BlockSize,
+				Count:     block.Count,
+				Iflag:     "fullblock",
+				Oflag:     "append",
+				Conv:      "notrunc",
```

Comment:
how about adding some comment for these different `Conv` value in different action

Reply:
OK.

Reply:
Fixed.